### ISSUE TITLE
multi gpu does not work with validation data

### ISSUE THREAD

**Issue by dl-nlp** [keras-team/keras#9404]
Hi 

trying to make multi gpu work with validation data. 
Tensorflow: 1.5
cudnn: 7
cuda: 9.0

Error: tensorflow/stream_executor/cuda/cuda_dnn.cc:444] could not convert BatchDescriptor {count: 0 feature_map_count: 32 spatial: 149 149  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX} to cudnn tensor descriptor: CUDNN_STATUS_BAD_PARAM
```
batch_size = 32
NUM_CLASSES = 2
train_datagen = ImageDataGenerator(...)
test_datagen = ImageDataGenerator(...)
train_generator = train_datagen.flow_from_directory(...)  
validation_generator = test_datagen.flow_from_directory(...)
base_model = InceptionV3(weights='imagenet', include_top=False)
x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
predictions = Dense(2, activation='softmax')(x)
model = Model(inputs=base_model.input, outputs=predictions)
model = multi_gpu_model(model,2)


model.compile(optimizer='rmsprop', loss='categorical_crossentropy' , metrics=['accuracy'])
print(model.summary())
model.fit_generator(train_generator, steps_per_epoch=2, epochs=10,                validation_data=validation_generator)
```

Error occurs when validation happens. Seems like it does not split validation data on both gpu -> one gpu gets an empty batch. Works fine on single gpu or on multiple without validation data.

Help appreciated!

**Comment by sulantha2006** [comment]
Was this solved?

**Comment by Pavel-Akapian** [comment]
I also faced quite similar problem with validation but not 100% sure yet about reasons of my problem. I get nans on validation or impossible values of loss on validation (train loss is absolutely ok).

**Comment by shsshs** [comment]
Any news on this?

**Comment by michaelmunie** [comment]
This is still a problem and referenced in #5406 #8358.
Option 1: you can also go the route of deleting data points from your data set so it is evenly dividable by your batch size.  (or at least such that n % batch_size > num_gpu)  The trouble with this is you need to re-evaluate and re-delete or un-delete if you change num_gpus or batch_size.
Option 2: Make the keras iterator enforce even batch sizes. I believe this is the better solution. Obviously, this will still throw an exception if one sets the batch size less than the number of gpus.  This also causes not all examples to be seen every batch, and causes the last few examples to never be seen if shuffle=False.


At your own risk, replacing the following methods with the code below in class Iterator in keras.preprocessing.image.py worked for me.
```
    def __init__(self, n, batch_size, shuffle, seed):
        self.n = n
        self.batch_size = batch_size
        self.seed = seed
        self.shuffle = shuffle
        self.batch_index = 0
        self.total_batches_seen = 0
        self.lock = threading.Lock()
        self.index_array = None
        self.index_generator = self._flow_index()
        self.excess_elements = self.n % self.batch_size

    def _set_index_array(self):
        self.index_array = np.arange(self.n)
        if self.shuffle:
            # self.index_array = np.random.permutation(self.n)
            self.index_array = np.random.permutation(self.n)[:self.n - self.excess_elements]
        else:
            self.index_array = self.index_array[: self.n - self.excess_elements]

    def __len__(self):
        #return (self.n + self.batch_size - 1) // self.batch_size  # round up
        return (self.n) // self.batch_size  # round down

    def _flow_index(self):
        # Ensure self.batch_index is 0.
        self.reset()
        while 1:
            if self.seed is not None:
                np.random.seed(self.seed + self.total_batches_seen)
            if self.batch_index == 0:
                self._set_index_array()

            current_index = (self.batch_index * self.batch_size) % self.n
            if self.n > current_index + self.batch_size + self.excess_elements:
                self.batch_index += 1
            else:
                self.batch_index = 0
            self.total_batches_seen += 1
            yield self.index_array[current_index:
                                   current_index + self.batch_size]
```
